README.md
# cupertino_notification_bar_ui

**Hi, welcome here!**

Myself ***Delwin***, a passionate `FlutterFire` Developer.

## Introduction

This is a package made for Flutter on Android, iOS, Windows and Linux.   
UI design for dismissible  `Containers` similar to that of iOS Notification Center. Inspired by a question on StackOverflow. 

<img src="https://user-images.githubusercontent.com/84124091/219048415-e6a90398-29ed-4da1-971c-7477b94d13fb.png" height=480 width=270>

## Features
You can organize UI components in a stacked manner. These are only dismissible and can trigger a `Function` on clicked. 

`StackedListItem` is a built in `StatelessWidget` which can be coupled with `StackedList` to show components. Alternatively you can use `Containers` for the same.


<video src="https://user-images.githubusercontent.com/84124091/219039940-beb32e6c-a7e3-4e9b-8f9a-4a6060b4d3ed.mp4" data-canonical-src="https://user-images.githubusercontent.com/84124091/219039940-beb32e6c-a7e3-4e9b-8f9a-4a6060b4d3ed.mp4" controls="controls" muted="muted" style="max-height:640px; min-height: 200px">
</video>

## Getting started
### Depend on it
Add package to `pubspec.yaml` file
```yaml
dependencies:
  cupertino_notification_bar_ui: <latest>
```
Alternatively, your editor might support  `flutter pub get`. Check the docs for your editor to learn more.

##### Or alternatively run 
```
flutter pub add cupertino_notification_bar_ui
```

### Import it
Now in your Dart code, you can use:   
```dart
import 'package:cupertino_notification_bar_ui/cupertino_notification_bar_ui.dart';
```

## Usage

  The package takes in a list of children, either of `StackedListItem` or `Container` as its parameter. 
Use the below given example to set up

```dart
import 'package:flutter/material.dart';
import 'package:cupertino_notification_bar_ui/cupertino_notification_bar_ui.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  
  // define `StackItems` here.
  // afternatively you can use Containers with height to 75px
  final stackItems = [
    StackedListItem(
      leading: const CircleAvatar(
          backgroundImage: NetworkImage(
              "https://cdn.pixabay.com/photo/2014/02/27/16/10/flowers-276014_960_720.jpg")),
      title: "Hi",
      subtitle: const Text("It's nice to see you here"),
      trailing: const Icon(Icons.link_rounded),
    ),
    StackedListItem(
      leading: const CircleAvatar(
          backgroundImage: NetworkImage(
              "https://cdn.pixabay.com/photo/2014/02/27/16/10/flowers-276014_960_720.jpg")),
      title: "This happens to be my first open-source",
      subtitle: const Text("We all start somewhere, right?"),
      trailing: const Icon(Icons.link_rounded),
    ),
    StackedListItem(
      leading: const CircleAvatar(
          backgroundImage: NetworkImage(
              "https://cdn.pixabay.com/photo/2014/02/27/16/10/flowers-276014_960_720.jpg")),
      title: "Thanks for checking this out!",
      subtitle: const Text("Happy coding :)"),
      trailing: const Icon(Icons.link_rounded),
    )
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: Colors.deepOrange.shade200,
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: StackedList(children: stackItems));
  }
}

```


## Additional information
Check => [official GitHub repo](https://github.com/opxdelwin/cupertino_notification_bar_ui).

This package was made just with the intention of providing a structured solution. I do not practice open source regularly, hence I will not be managing this package actively. 

Feel free to raise a issue in repository for this package on GitHub, I'd love to tailor this package as per my availability, or we could work together on this.

If you really liked this package, even though I do not actively want to be sponsored, you can donate here.    
<img src="https://user-images.githubusercontent.com/84124091/219038251-9749d34f-8f8e-4be4-af44-7868dd7e6aca.png" height=240 width=240>


CHANGELOG.md
## 1.0.0
Initial Release

## 1.0.1
Metadata updates

## 1.0.101
videos and photos added in README.md

lib/cupertino_notification_bar_ui.dart
library cupertino_notification_bar_ui;

import "package:flutter/material.dart";

class StackedList extends StatefulWidget {
  StackedList({
    required this.children,
    this.dismissDirection = DismissDirection.startToEnd,
    this.verticalDelta = 13,
    this.scaleFactor = 20,
    this.animationDuration = 200,
    super.key,
  }) : assert(children.isNotEmpty);

  /// List containing all the widgets to be showed as part of stacked UI
  /// this cannot be null
  final List<Widget> children;

  /// Vertical factor for gap between `StackedListItem`
  /// defaults to 13. Higher the value, more the gap.
  final double verticalDelta;

  /// Scale factor for undelying `StackedListItem`
  /// defaults to 20. More the value, more it scales down.
  final double scaleFactor;

  /// Direction of dismiss, defaults to Left to Right (`startToEnd`)
  final DismissDirection dismissDirection;

  /// New item animation duration
  final int animationDuration;

  @override
  State<StackedList> createState() => _StackedListState();
}

class _StackedListState extends State<StackedList> {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: Stack(
          alignment: Alignment.center,
          children: widget.children
              .map((element) {
                final index = widget.children.indexOf(element);
                return AnimatedPositioned(
                  duration: Duration(milliseconds: widget.animationDuration),
                  top: 200 + (index * widget.verticalDelta),
                  left: 0,
                  right: 0,
                  child: AnimatedScale(
                    duration:
                        Duration(milliseconds: widget.animationDuration + 50),
                    scale: index <= 4
                        ? 1 - (index * index / widget.scaleFactor)
                        : 0,
                    child: Dismissible(
                        direction: widget.dismissDirection,
                        key: UniqueKey(),
                        onDismissed: (DismissDirection direction) {
                          return setState(() {
                            widget.children.removeAt(index);
                          });
                        },
                        child: element),
                  ),
                );
              })
              .toList()
              //using reversed list as Stack works on Last-Come-First-Out
              .reversed
              .toList()),
    );
  }
}

class StackedListItem extends StatelessWidget {
  // ignore: prefer_const_constructors_in_immutables
  StackedListItem({
    super.key,
    required this.title,
    this.leading,
    this.onTapFunction,
    this.titleStyle,
    this.trailing,
    this.subtitle,
    this.isThreeLine = false,
    this.borderWidth = 2,
    this.backgroundColor = Colors.white,
    this.borderColor = Colors.white30,
  });

  /// leading image
  final Widget? leading;

  /// set this to true if you will be using subtitle, defaults to false
  final bool isThreeLine;

  /// callback function, defaults null
  final void Function()? onTapFunction;

  /// title or main content of tile
  final String title;

  /// styling for title, defaults to null
  final TextStyle? titleStyle;

  /// trailing content
  final Widget? trailing;

  /// subtitle widget, set `isThreeLine` to true
  final Widget? subtitle;

  /// background color for the widget
  final Color backgroundColor;

  final Color borderColor;
  final double borderWidth;

  @override
  Widget build(BuildContext context) {
    return Container(
      height: 75,
      decoration: BoxDecoration(
        color: backgroundColor,
        borderRadius: BorderRadius.circular(8),
        border: Border.all(width: borderWidth, color: borderColor),
      ),
      child: ListTile(
        leading: leading,
        isThreeLine: isThreeLine,
        onTap: onTapFunction,
        title: Text(
          title,
          style: titleStyle,
        ),
        trailing: trailing,
        subtitle: subtitle,
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:cupertino_notification_bar_ui/cupertino_notification_bar_ui.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  // define `StackItems` here.
  // afternatively you can use Containers with height to 75px
  final stackItems = [
    StackedListItem(
      leading: const CircleAvatar(
          backgroundImage: NetworkImage(
              "https://cdn.pixabay.com/photo/2014/02/27/16/10/flowers-276014_960_720.jpg")),
      title: "Hi",
      subtitle: const Text("It's nice to see you here"),
      trailing: const Icon(Icons.link_rounded),
    ),
    StackedListItem(
      leading: const CircleAvatar(
          backgroundImage: NetworkImage(
              "https://cdn.pixabay.com/photo/2014/02/27/16/10/flowers-276014_960_720.jpg")),
      title: "This happens to be my first open-source",
      subtitle: const Text("We all start somewhere, right?"),
      trailing: const Icon(Icons.link_rounded),
    ),
    StackedListItem(
      leading: const CircleAvatar(
          backgroundImage: NetworkImage(
              "https://cdn.pixabay.com/photo/2014/02/27/16/10/flowers-276014_960_720.jpg")),
      title: "Thanks for checking this out!",
      subtitle: const Text("Happy coding :)"),
      trailing: const Icon(Icons.link_rounded),
    )
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: Colors.deepOrange.shade200,
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: StackedList(children: stackItems));
  }
}
